taichi/backends/vulkan/codegen_vulkan.cpp

```diff
@@ -257,208 +312,365 @@ class TaskCodegen : public IRVisitor {
   void visit(GlobalStoreStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
     const auto dt = stmt->val->element_type();
-    emit("{} = {};", at_buffer(stmt->dest, dt),
-         store_as_int_bits(stmt->val->raw_name(), dt));
+    spirv::Value buffer_ptr; 
+    spirv::Value val = ir_->query_value(stmt->val->raw_name());
+    if (ptr_to_buffers_.at(stmt->dest) == BuffersEnum::Root) {
+      buffer_ptr = ir_->query_value(stmt->dest->raw_name());
+      buffer_ptr.flag = spirv::ValueKind::kVariablePtr; // make this value could store/load
+    } else {
+      buffer_ptr = at_buffer(stmt->dest, dt);
+    }
+    ir_->store_variable(buffer_ptr, ir_->make_value(spv::OpBitcast, ir_->i32_type(), val));
   }
 
   void visit(GlobalLoadStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
     auto dt = stmt->element_type();
-    const auto loaded_int = at_buffer(stmt->src, dt);
-    emit("const {} {} = {};", opengl_data_type_name(dt), stmt->raw_name(),
-         load_from_int_bits(loaded_int, dt));
+    spirv::Value buffer_ptr;
+    spirv::Value val;
+    if (ptr_to_buffers_.at(stmt->src) == BuffersEnum::Root) {
+      buffer_ptr = ir_->query_value(stmt->src->raw_name());
+      buffer_ptr.flag = spirv::ValueKind::kVariablePtr; // make this value could store/load
+    } else {
+      buffer_ptr = at_buffer(stmt->src, dt);
+    }
+    val = ir_->make_value(spv::OpBitcast, ir_->get_primitive_type(dt),
+          ir_->load_variable(buffer_ptr, ir_->i32_type()));
+    ir_->register_value(stmt->raw_name(), val);
   }
 
   void visit(ArgLoadStmt *stmt) override {
     const auto arg_id = stmt->arg_id;
     const auto &arg_attribs = ctx_attribs_->args()[arg_id];
     const auto offset_in_mem = arg_attribs.offset_in_mem;
     if (stmt->is_ptr) {
-      emit("// Pointer arg: id={} offset_in_mem={}", arg_id, offset_in_mem);
       // Do not shift! We are indexing the buffers at byte granularity.
-      emit("const int {} = {};", stmt->raw_name(), offset_in_mem);
+      spirv::Value val = ir_->int_immediate_number(ir_->i32_type(), offset_in_mem);
+      ir_->register_value(stmt->raw_name(), val);
```

Comment:
nit: Maybe look into Vulkan's unique feature: PushConstants. PushConstants should be perfect for arguments (that is written to / updated almost every dispatch call, and small in size).